src/corehost/cli/ijwhost/ijwthunk.cpp

```diff
@@ -0,0 +1,189 @@
+
+    {
+        std::lock_guard<std::mutex> _(g_thunkChunkLock);
+        chunk->SetNext(g_pVtableBootstrapThunkChunkList);
+        g_pVtableBootstrapThunkChunkList = chunk;
+    }
+
+    size_t currentThunk = 0;
+    for(size_t i = 0; i < numFixupRecords; ++i)
+    {
+#ifndef _WIN64
+        assert((pFixupTable[i].Type & (COR_VTABLE_FROM_UNMANAGED | COR_VTABLE_FROM_UNMANAGED_RETAIN_APPDOMAIN)) && "managed->managed vtablefixup slots are not supported!");
+#endif
+        if (pFixupTable[i].Type & COR_VTABLE_PTRSIZED)
+        {
+            const BYTE** pointers = (const BYTE**)pe.GetRvaData(pFixupTable[i].RVA);
+
+#ifdef _WIN64
+            DWORD oldProtect;
+            if(!VirtualProtect(pointers, (sizeof(BYTE*) * pFixupTable[i].Count), PAGE_READWRITE, &oldProtect))
+            {
+                trace::error(_X("Failed to change the vtfixup table from RO to R/W failed.\n"));
+                return false;
+            }
+#endif
+
+            for (std::uint16_t method = 0; method < pFixupTable[i].Count; method++)
+            {
+                mdToken tok = (mdToken)(std::uintptr_t) pointers[method];
+                assert (TypeFromToken(tok) == mdtMethodDef || TypeFromToken(tok) == mdtMemberRef);
+                bootstrap_thunk* pThunk = chunk->GetThunk(currentThunk++);
+                pThunk->initialize((std::uintptr_t)&start_runtime_thunk_stub,
+                                    (pal::dll_t)pe.GetBase(),
+                                    tok,
+                                    (std::uintptr_t *)&pointers[method]);
+                pointers[method] = (BYTE*)pThunk->get_entrypoint();
+            }
+
+#ifdef _WIN64
+            DWORD _;
+            if(!VirtualProtect(pointers, (sizeof(BYTE*) * pFixupTable[i].Count), oldProtect, &_))
+            {
+                trace::warning(_X("Failed to change the vtfixup table from R/W back to RO failed.\n"));
+            }
+#endif
+        }
+    }
+
+    return true;
+}
+
+extern "C" std::uintptr_t __stdcall start_runtime_and_get_target_address(std::uintptr_t cookie)
+{
+    bootstrap_thunk *pThunk = bootstrap_thunk::get_thunk_from_cookie(cookie);
+
+    load_in_memory_assembly_fn loadInMemoryAssembly;
+    pal::dll_t moduleHandle = pThunk->get_dll_handle();
+    pal::hresult_t status = get_load_in_memory_assembly_delegate(moduleHandle, &loadInMemoryAssembly);
+
+    if (status != StatusCode::Success)
+    {
+        // If we ignore the failure to patch bootstrap thunks we will come to this same
+        // function again, causing an infinite loop of "Failed to start the .NET Core runtime" errors.
+        // As we were taken here via an entry point with arbitrary signature,
+        // there's no way of returning the error code so we just throw it.
+
+        trace::error(_X("Failed to start the .NET Core runtime. Error code %d"), status);
+
+#pragma warning (push)
+#pragma warning (disable: 4297)
+        throw status;
```

Comment:
This is a pure C++ exception. It's the same behavior as .NET Framework in this case.

Reply:
This call also doesn't happen during load. This is in the delayed-activation thunk, so the code here is called by whatever method first calls a managed method in the IJW host. As a result, we have absolutely no idea what code called us, so there's no good way to report the errors to the caller since the caller can be literally anything.

Reply:
I understand... 
I guess we picked C++ since this is most likely called by the native parts of the C++/CLI and thus something which at least resembles C++ and probably has catch handlers for it... but I'm just guessing.

Anyway - it is a conscious choice with history behind it - good enough for me :-)